c_src/enif_jq.c

```diff
@@ -190,17 +439,106 @@ static ERL_NIF_TERM parse_nif(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[
 
 static ErlNifFunc nif_funcs[] = {
     /*
-       The parse_nif function seems to be very slow.
+       The parse_nif function seems to be very slow (at least when
+       given filter programs that are not in the cache).
        The Erlang VM acts very strangely when it is not scheduled
        on a dirty scheduler (for example, timer:sleep() suspends
        for a much longer time than is should).
     */
     {"parse", 2, parse_nif, ERL_NIF_DIRTY_JOB_CPU_BOUND}
 };
 
-int upgrade(ErlNifEnv* env, void** priv_data, void** old_priv_data, ERL_NIF_TERM load_info)
-{
-	return 0;
+static int get_int_config(
+        ErlNifEnv* caller_env,
+        ERL_NIF_TERM load_info,
+        char* property_name,
+        int* res_wb) {
+
+    ERL_NIF_TERM property_atom = enif_make_atom(
+            caller_env,
+            property_name);
+    ERL_NIF_TERM property_term; 
+    if ( ! enif_get_map_value(
+            caller_env,
+            load_info,
+            property_atom,
+            &property_term)) {
+        // Incorrect load info
+        return 1;
+    }
+    if ( ! enif_get_int(
+               caller_env,
+               property_term,
+               res_wb)) {
+        // Incorrect load info
+        return 1;
+    }
+    return 0;
+}
+
+static int load_helper(
+        ErlNifEnv* caller_env,
+        void** priv_data,
+        ERL_NIF_TERM load_info,
+        int nr_of_loads_before) {
+
+    int filter_program_lru_cache_max_size;
+    if ( get_int_config(
+                caller_env,
+                load_info,
+                "filter_program_lru_cache_max_size",
+                &filter_program_lru_cache_max_size) ) {
+        return 1;
+    }
+    module_private_data* data = enif_alloc(sizeof(module_private_data));
+    data->nr_of_loads_before = data->nr_of_loads_before + 1;
```

Comment:
Thanks for bringing up this issue. We could handle memory allocation errors in our code with some effort (e.g., long jumping to a place where we can give an error return). But the jq library calls functions that allocate memory without checking for errors in several locations (see https://github.com/terry-xiaoyu/jqc/blob/51df83ccf4808867eee79a011c71a93a121be1b6/src/compile.c#L158 for example), so we would not be safe if we don't also fix these locations. Additionally, I don't think the Erlang VM can handle a situation in which memory allocation fails in any other way than shutting down the VM, so I'm not sure if it's worth putting a lot of effort into handling memory allocation failure.

What do you think?

Reply:
I think `enif_alloc` could return NULL when the total VM memory usage is limited by some VM args and also it is nice to have Erlang crash dump rather than core dump.

We had offline chat about this and decide to just return some error without cleaning here.

Reply:
I have added code for handing when enif_alloc returns NULL in this commit [d9e4279](https://github.com/emqx/jq/pull/14/commits/d9e4279a0791030bf520c6de9720993d51d951f4)